### PR TITLE
monitor system-wide memory metrics

### PR DESCRIPTION
**Pull request by nagworld9**
## Description
Created new PeriodicOperation in Monitor thread to monitor system-wide available and used memory metrics. Those metrics reported every hour.

2022-06-13T22:03:28.923508Z INFO MonitorHandler ExtHandler Event: name=WALinuxAgent, op=METRIC, message=Metric Memory/Used MBytes [] = 449.234375, duration=0
2022-06-13T22:03:28.924656Z INFO MonitorHandler ExtHandler Event: name=WALinuxAgent, op=METRIC, message=Metric Memory/Available MBytes [] = 7212.453125, duration=0

Issue # 

---

### PR information
- [ ] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [ ] If applicable, the PR references the bug/issue that it fixes in the description.
- [ ] New Unit tests were added for the changes made

### Quality of Code and Contribution Guidelines
- [ ] I have read the [contribution guidelines](https://github.com/Azure/WALinuxAgent/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: azurelinuxagent/common/conf.py [hunk 1]
```diff
@@ -137,7 +137,8 @@ def load_conf_from_file(conf_file_path, conf=__conf__):
     "Debug.CgroupDisableOnProcessCheckFailure": True,
     "Debug.CgroupDisableOnQuotaCheckFailure": True,
     "Debug.EnableFastTrack": True,
-    "Debug.EnableGAVersioning": False
+    "Debug.EnableGAVersioning": False,
+    "Debug.LogSystemWideMetrics": False
```
Comment (narrieta) [inline azurelinuxagent/common/conf.py:141]: let's try to minimize those debug options... is this one really needed or can you just use CgroupLogMetrics?

File: azurelinuxagent/common/osutil/default.py [hunk 2]
```diff
@@ -1411,6 +1412,30 @@ def get_total_cpu_ticks_since_boot():
...
+        free_cmd = ["free", "-b"]
+        try:
+            memory = shellutil.run_command(free_cmd)
+            for line in memory.split("\n"):
+                if ALL_MEMS_REGEX.match(line):
+                    mems = line.split()
+                    used_mem = int(mems[2])
+                    available_mem = int(mems[6])  # see "man free" for a description of these fields
+        except CommandError as e:
+            logger.warn("Cannot get the memory table. {0} failed: {1}", ustr(free_cmd), ustr(e))
```
Comment (narrieta) [inline azurelinuxagent/common/osutil/default.py:1436]: if this fails continuously we would be producing a warning on each period

you can let the exception bubble up and handle it on the caller (the periodic task), which can log the error once (until it succeeds again).

File: azurelinuxagent/common/osutil/default.py [hunk 3]
```diff
@@ -1411,6 +1412,30 @@ def get_total_cpu_ticks_since_boot():
...
+        """
+        used_mem = available_mem = 0
+        free_cmd = ["free", "-b"]
+        try:
+            memory = shellutil.run_command(free_cmd)
+            for line in memory.split("\n"):
+                if ALL_MEMS_REGEX.match(line):
+                    mems = line.split()
+                    used_mem = int(mems[2])
+                    available_mem = int(mems[6])  # see "man free" for a description of these fields
```
Comment (narrieta) [inline azurelinuxagent/common/osutil/default.py:1434]: available would be 7

Comment (nagworld9) [reply azurelinuxagent/common/osutil/default.py:1434]: yes, that's why I used index `'6'`to get that.

Note: List start index is '`0`'

Comment (narrieta) [reply azurelinuxagent/common/osutil/default.py:1434]: 0 is the comment, total is 2, available 7

Comment (narrieta) [reply azurelinuxagent/common/osutil/default.py:1434]: or you added the comment and total should be 1?

Comment (narrieta) [reply azurelinuxagent/common/osutil/default.py:1434]: oh, used, not total --- never mind, you are ok

**Review by narrieta — APPROVED** [review]

**Pull request merged**